File: docs/chapter1/section3/src/4-3_add.go
package main

import (
	"net/http"

	"github.com/labstack/echo/v4"
)

// Jsonで受け取って構造体に変換するための構造体を定義
type requestData struct {
	Left  *int `json:"left"`
	Right *int `json:"right,omitempty"`
}

// 計算結果をJsonで返すための構造体を定義
type responseData struct {
	Answer int `json:"answer"`
}

// エラーをJsonで返すための構造体を定義
type errorData struct {
	Error string `json:"error,omitempty"`
}

func main() {
	e := echo.New()

	// 「/add」というエンドポイントを設定する
	e.POST("/add", addHandler)

	e.Logger.Fatal(e.Start(":8080"))
}

func addHandler(c echo.Context) error {
	// 受け取りたい JSON を示す空の変数を先に用意する。
	numbers := &requestData{}
	// 受け取った JSON を data に代入する
	err := c.Bind(numbers)

	// JSON を data に代入する時にエラーが発生したり、値が空だった時、つまりリクエストが適切ではなかった時、
	if err != nil || numbers.Left == nil || numbers.Right == nil {
		// ステータスコード 400 Bad Request で"Bad Request"が入ったJSONを返す
		return c.JSON(http.StatusBadRequest, &errorData{Error: "Bad Request"})
	}

	// 計算結果
	result := *numbers.Left + *numbers.Right
	// 計算結果をステータスコード 200 Ok で返す
	return c.JSON(http.StatusOK, &responseData{Answer: result})
}

File: docs/chapter1/section3/src/4-3_add.rs
use axum::extract::rejection::JsonRejection;
use axum::{http::StatusCode, routing::post, Json, Router};

#[tokio::main]
async fn main() {
    // 「/ping」というエンドポイントを設定する
    let app = Router::new().route("/add", post(add_handler));

    // ポート8080でリスナーを作成する
    let listener = tokio::net::TcpListener::bind("127.0.0.1:8080")
        .await
        .unwrap();

    println!("listening on {}", listener.local_addr().unwrap());

    // サーバーを起動する
    axum::serve(listener, app).await.unwrap();
}

// クエリパラメータを受け取るための構造体を定義
#[derive(serde::Deserialize)]
struct AddQuery {
    left: f64,
    right: f64,
}

// レスポンスとして返す構造体を定義
#[derive(serde::Serialize)]
struct AddResponse {
    result: i64,
}

// エラーレスポンスとして返す構造体を定義
#[derive(serde::Serialize)]
struct AddError {
    error: String,
}

async fn add_handler(
    query: Result<Json<AddQuery>, JsonRejection>,
) -> Result<Json<AddResponse>, (StatusCode, Json<AddError>)> {
    match query {
        // クエリが正しく受け取れた場合、クライアントに結果を返す
        Ok(query) => Ok(Json(AddResponse {
            result: (query.left + query.right) as i64,
        })),
        // クエリが正しく受け取れなかった場合、エラーを返す
        Err(_) => Err((
            StatusCode::BAD_REQUEST,
            Json(AddError {
                error: String::from("Bad Request"),
            }),
        )),
    }
}